newton/core/model.py

```diff
@@ -446,137 +381,109 @@ def control(self, requires_grad: bool | None = None, clone_variables: bool = Tru
-            target.soft_contact_body_vel = wp.zeros(count, dtype=wp.vec3, requires_grad=requires_grad)
-            target.soft_contact_normal = wp.zeros(count, dtype=wp.vec3, requires_grad=requires_grad)
-            target.soft_contact_tids = wp.zeros(self.particle_count * (self.shape_count - 1), dtype=int)
+        # Get the body index for this shape
+        body_idx = shape_body[shape_idx]
 
-    def allocate_soft_contacts(self, count, requires_grad=False):
-        self._allocate_soft_contacts(self, count, requires_grad)
+        # If shape is attached to a body (body_idx >= 0), concatenate transforms
+        if body_idx >= 0:
+            # Get the body transform in world space
+            X_wb = body_q[body_idx]
 
-    def count_contact_points(self):
-        """
-        Counts the maximum number of rigid contact points that need to be allocated.
-        This function returns two values corresponding to the maximum number of potential contacts
-        excluding the limiting from `Model.rigid_mesh_contact_max` and the maximum number of
-        contact points that may be generated when considering the `Model.rigid_mesh_contact_max` limit.
-
-        :returns:
-            - potential_count (int): Potential number of contact points
-            - actual_count (int): Actual number of contact points
+            # Concatenate: world_transform = body_transform * shape_transform
+            X_ws = wp.transform_multiply(X_wb, X_bs)
+            shape_world_transform[shape_idx] = X_ws
+        else:
+            # Shape is not attached to a body (static shape), use local
+            # transform as world transform
+            shape_world_transform[shape_idx] = X_bs
+
+    def collide(
+        self: Model,
+        state: State,
+        requires_grad: bool | None = None,
+        edge_sdf_iter: int = 10,
+        iterate_mesh_vertices: bool = True,
+        max_contacts_per_pair: int = 10,
+        soft_contact_margin: float = 0.0,
+        rigid_contact_margin: float = 0.0,
+    ) -> Contact:
+        """Generate contact points for the particles and rigid bodies in the
+        model for use in contact-dynamics kernels.
+
+        Args:
+            state: The state of the model.
+            requires_grad: Whether to duplicate contact arrays for gradient
+                computation (if ``None``, uses ``self.requires_grad``).
+            edge_sdf_iter: Number of search iterations for finding closest
+                contact points between edges and SDF.
+            iterate_mesh_vertices: Whether to iterate over all vertices of a
+                mesh for contact generation (used for capsule/box <> mesh
+                collision).
+            max_contacts_per_pair: Maximum number of contacts per shape pair.
+            soft_contact_margin: Margin for soft contact generation.
+            rigid_contact_margin: Margin for rigid contact generation.
```

Comment:
I think these `CollisionPipeline` settings should be removed here, we should have defaults for those in the `CollisionPipeline` itself. Instead, there should only be an optional `CollisionPipeline` object the user can pass in here:

```py
def collide(
        self: Model,
        state: State,
        collision_pipeline: CollisionPipeline | None = None,
) -> Contacts:
       ....
```